File: kaji/packages/ts/src/tasks/handle.ts
import { InMemoryEventCommitter } from "@/events/committer";
import type { EventCommitter } from "@/events/protocols";
import { InMemoryEventStore, type EventStore } from "@/events/store";
import {
  TaskCancelled,
  TaskCreated,
  TaskResumed,
  ToolApprovalApproved,
  ToolApprovalRejected,
  type NewKajiEvent,
  type StoredKajiEvent,
} from "@/events/schemas";
import { systemIdFactory, type IdFactory } from "@/internal/uuid";
import {
  InMemorySessionTurnCoordinator,
  type SessionTurnCoordinator,
} from "@/runtime/session/coordinator";

import { approvalId, projectTask } from "./projector";
import type { PendingApproval, TaskSnapshot } from "./types";

export class TaskNotFoundError extends Error {}

export class InMemoryBackend {
  readonly taskSessions = new Map<string, string>();

  constructor(
    readonly store: EventStore = new InMemoryEventStore(),
    readonly journal: EventCommitter = new InMemoryEventCommitter(store),
    readonly coordinator: SessionTurnCoordinator = new InMemorySessionTurnCoordinator(),
  ) {}
}

export class TaskRuntime {
  constructor(
    private readonly backend: InMemoryBackend,
    private readonly ids: IdFactory = systemIdFactory,
  ) {}

  static forInMemory(options: { ids?: IdFactory } = {}): TaskRuntime {
    return new TaskRuntime(new InMemoryBackend(), options.ids);
  }

  async start(options: {
    session_id: string;
    principal_id: string;
    input: string;
    metadata?: Record<string, unknown>;
    task_id?: string;
  }): Promise<TaskHandle> {
    const taskId = options.task_id ?? this.ids.next("task");
    if (this.backend.taskSessions.has(taskId)) throw new Error(`task already exists: ${taskId}`);
    const handle = TaskHandle.forInMemory(taskId, options.session_id, this.backend, this.ids);
    await this.backend.coordinator.runExclusive(options.session_id, undefined, async () => {
      await this.backend.journal.commit(
        TaskCreated.parse({
          type: "task.created",
          id: this.ids.next("event"),
          task_id: taskId,
          session_id: options.session_id,
          principal_id: options.principal_id,
          input: options.input,
          metadata: options.metadata ?? {},
        }),
      );
      await this.backend.journal.commit(
        TaskResumed.parse({
          type: "task.resumed",
          id: this.ids.next("event"),
          task_id: taskId,
          session_id: options.session_id,
        }),
      );
      this.backend.taskSessions.set(taskId, options.session_id);
    });
    return handle;
  }

  get(taskId: string): TaskHandle {
    const sessionId = this.backend.taskSessions.get(taskId);
    if (!sessionId) throw new TaskNotFoundError(`unknown task: ${taskId}`);
    return TaskHandle.forInMemory(taskId, sessionId, this.backend, this.ids);
  }
}

export class TaskHandle {
  private constructor(
    readonly taskId: string,
    readonly sessionId: string,
    private readonly backend: InMemoryBackend,
    private readonly ids: IdFactory,
  ) {}

  static forInMemory(
    taskId: string,
    sessionId: string,
    backend: InMemoryBackend,
    ids: IdFactory = systemIdFactory,
  ): TaskHandle {
    return new TaskHandle(taskId, sessionId, backend, ids);
  }

  async events(options: { after_sequence?: number } = {}): Promise<readonly StoredKajiEvent[]> {
    return this.backend.store.getEvents(this.sessionId, {
      afterSequence: options.after_sequence ?? 0,
    });
  }

  async snapshot(): Promise<TaskSnapshot> {
    return projectTask(this.taskId, await this.events());
  }

  async pendingApprovals(): Promise<readonly PendingApproval[]> {
    const snapshot = await this.snapshot();
    const pending = new Map<string, PendingApproval>();
    for (const event of await this.events()) {
      if (event.type === "tool.approval.requested") {
        const id = approvalId(event);
        pending.set(id, {
          id,
          capability: event.tool_name,
          risk: event.risk,
          arguments: event.tool_args,
        });
      }
    }
    return snapshot.pending_approvals.map((item) => pending.get(item.id)!).filter(Boolean);
  }

  async cancel(): Promise<TaskSnapshot> {
    return this.append(
      TaskCancelled.parse({
        type: "task.cancelled",
        id: this.ids.next("event"),
        task_id: this.taskId,
        session_id: this.sessionId,
      }),
    );
  }

  async resume(): Promise<TaskSnapshot> {
    return this.append(
      TaskResumed.parse({
        type: "task.resumed",
        id: this.ids.next("event"),
        task_id: this.taskId,
        session_id: this.sessionId,
      }),
    );
  }

  async decideApproval(
    approval: PendingApproval,
    approved: boolean,
    reason = "host decision",
  ): Promise<TaskSnapshot> {
    const [turn_id, tool_call_id] = approval.id.split(":", 2);
    const base = {
      id: this.ids.next("event"),
      session_id: this.sessionId,
      turn_id,
      tool_call_id,
      tool_name: approval.capability,
    };
    return this.append(
      approved
        ? ToolApprovalApproved.parse({ type: "tool.approval.approved", ...base })
        : ToolApprovalRejected.parse({
            type: "tool.approval.rejected",
            ...base,
            error_code: "APPROVAL_REJECTED",
            reason,
          }),
    );
  }

  private async append(event: NewKajiEvent): Promise<TaskSnapshot> {
    return this.backend.coordinator.runExclusive(this.sessionId, undefined, async () => {
      await this.backend.journal.commit(event);
      return projectTask(this.taskId, await this.events());
    });
  }
}

File: kaji/packages/py/src/tasks/handle.py
"""In-memory journal-backed task handles."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from kaji.core.determinism import IdFactory, SYSTEM_ID_FACTORY
from kaji.events import InMemoryEventJournal, InMemoryEventStore
from kaji.events.protocols import EventJournal
from kaji.events.schemas import (
    NewKajiEvent,
    TaskCancelled,
    TaskCreated,
    TaskResumed,
    StoredKajiEvent,
    ToolApprovalApproved,
    ToolApprovalRejected,
)
from kaji.events.store import EventStore
from kaji.runtime.agents.coordinator import (
    TurnCoordinator,
    default_coordinator_for_store,
)
from kaji.runtime.tools.idempotency import (
    InMemoryToolIdempotencyLedger,
    ToolIdempotencyLedger,
)

from .errors import TaskNotFoundError
from .projector import approval_id, project_task
from .types import PendingApproval, TaskSnapshot


@dataclass(slots=True)
class InMemoryBackend:
    """Minimal composition root for the task prototype."""

    store: EventStore
    journal: EventJournal
    coordinator: TurnCoordinator
    idempotency_ledger: ToolIdempotencyLedger = field(
        default_factory=InMemoryToolIdempotencyLedger
    )
    task_sessions: dict[str, str] = field(default_factory=dict)

    @classmethod
    def create(cls) -> InMemoryBackend:
        store = InMemoryEventStore()
        return cls(
            store, InMemoryEventJournal(store), default_coordinator_for_store(store)
        )


class TaskRuntime:
    """Creates and finds task handles; snapshots always replay the journal."""

    def __init__(
        self, backend: InMemoryBackend, *, ids: IdFactory = SYSTEM_ID_FACTORY
    ) -> None:
        self._backend = backend
        self._ids = ids

    @classmethod
    def for_in_memory(cls, *, ids: IdFactory = SYSTEM_ID_FACTORY) -> TaskRuntime:
        return cls(InMemoryBackend.create(), ids=ids)

    async def start(
        self,
        *,
        session_id: str,
        principal_id: str,
        input: str,
        metadata: dict[str, Any] | None = None,
        task_id: str | None = None,
    ) -> TaskHandle:
        identifier = task_id or self._ids.next("task")
        if identifier in self._backend.task_sessions:
            raise ValueError(f"task already exists: {identifier}")
        handle = TaskHandle.for_in_memory(
            identifier, session_id, self._backend, self._ids
        )
        async with self._backend.coordinator.acquire(session_id):
            await self._backend.journal.commit(
                TaskCreated(
                    id=self._ids.next("event"),
                    task_id=identifier,
                    session_id=session_id,
                    principal_id=principal_id,
                    input=input,
                    metadata=metadata or {},
                )
            )
            await self._backend.journal.commit(
                TaskResumed(
                    id=self._ids.next("event"),
                    task_id=identifier,
                    session_id=session_id,
                )
            )
            self._backend.task_sessions[identifier] = session_id
        return handle

    def get(self, task_id: str) -> TaskHandle:
        try:
            session_id = self._backend.task_sessions[task_id]
        except KeyError as exc:
            raise TaskNotFoundError(f"unknown task: {task_id}") from exc
        return TaskHandle.for_in_memory(task_id, session_id, self._backend, self._ids)


class TaskHandle:
    """A narrow task control surface over one session journal."""

    def __init__(
        self, task_id: str, session_id: str, backend: InMemoryBackend, ids: IdFactory
    ) -> None:
        self.task_id = task_id
        self.session_id = session_id
        self._backend = backend
        self._ids = ids

    @classmethod
    def for_in_memory(
        cls,
        task_id: str,
        session_id: str,
        backend: InMemoryBackend,
        ids: IdFactory = SYSTEM_ID_FACTORY,
    ) -> TaskHandle:
        return cls(task_id, session_id, backend, ids)

    async def events(self, *, after_sequence: int = 0) -> tuple[StoredKajiEvent, ...]:
        return tuple(
            await self._backend.store.get_events(
                self.session_id, after_sequence=after_sequence
            )
        )

    async def snapshot(self) -> TaskSnapshot:
        return project_task(self.task_id, await self.events())

    async def pending_approvals(self) -> tuple[PendingApproval, ...]:
        events = await self.events()
        snapshot = project_task(self.task_id, events)
        by_id: dict[str, PendingApproval] = {}
        from .types import PendingApproval as FullApproval

        for event in events:
            if event.type.value == "tool.approval.requested":
                by_id[approval_id(event)] = FullApproval(
                    id=approval_id(event),
                    capability=event.tool_name,
                    risk=event.risk,
                    arguments=dict(event.tool_args),
                )
        return tuple(by_id[item.id] for item in snapshot.pending_approvals)

    async def cancel(self) -> TaskSnapshot:
        return await self._append(
            TaskCancelled(
                id=self._ids.next("event"),
                task_id=self.task_id,
                session_id=self.session_id,
            )
        )

    async def resume(self) -> TaskSnapshot:
        return await self._append(
            TaskResumed(
                id=self._ids.next("event"),
                task_id=self.task_id,
                session_id=self.session_id,
            )
        )

    async def decide_approval(
        self,
        approval: PendingApproval,
        *,
        approved: bool,
        reason: str = "host decision",
    ) -> TaskSnapshot:
        event: NewKajiEvent
        event_id = self._ids.next("event")
        turn_id, tool_call_id = approval.id.split(":", 1)
        if approved:
            event = ToolApprovalApproved(
                id=event_id,
                session_id=self.session_id,
                turn_id=turn_id,
                tool_call_id=tool_call_id,
                tool_name=approval.capability,
            )
        else:
            event = ToolApprovalRejected(
                id=event_id,
                session_id=self.session_id,
                turn_id=turn_id,
                tool_call_id=tool_call_id,
                tool_name=approval.capability,
                error_code="APPROVAL_REJECTED",
                reason=reason,
            )
        # The active EventApprovalHandler holds the session turn lease while it
        # waits, so this canonical decision must not queue behind that lease.
        await self._backend.journal.commit(event)
        return project_task(self.task_id, await self.events())

    async def _append(self, event: NewKajiEvent) -> TaskSnapshot:
        async with self._backend.coordinator.acquire(self.session_id):
            await self._backend.journal.commit(event)
            return project_task(self.task_id, await self.events())
